File: Advent of Code/Challenge Solutions/Year 2025/ChallengeSolution01.cs
// Task: https://adventofcode.com/2025/day/1

using Advent_of_Code.Utilities;

namespace Advent_of_Code.Challenge_Solutions.Year_2025;

public class ChallengeSolution01(IConsole console, ISolutionReader<ChallengeSolution01> reader)
    : ChallengeSolution<ChallengeSolution01>(console, reader)
{
    private const int UpperDialLimit = 100;
    private const int StartingDialPosition = 50;
    private const int Zero = 0;

    public override void SolveFirstPart()
    {
        var rotations = ReadRotations();
        var password = 0;
        var currentDialPosition = StartingDialPosition;

        foreach (var rotation in rotations)
        {
            currentDialPosition = TurnDial(currentDialPosition, rotation);

            if (currentDialPosition == Zero)
            {
                password++;
            }
        }

        _console.WriteLine($"Password: {password}");
    }

    public override void SolveSecondPart()
    {
        var rotations = ReadRotations();
        var password = 0;
        var currentDialPosition = StartingDialPosition;

        foreach (var rotation in rotations)
        {
            var fullRotations = rotation.Distance / UpperDialLimit;
            var remainder = rotation.Distance - (fullRotations * UpperDialLimit);

            var hasDialMovedPastZero = currentDialPosition != Zero
                && rotation.Direction switch
                {
                    Direction.Left => currentDialPosition <= remainder,
                    Direction.Right => currentDialPosition >= UpperDialLimit - remainder,
                    _ => throw new NotImplementedException()
                };

            password += fullRotations + (hasDialMovedPastZero ? 1 : 0);

            currentDialPosition = TurnDial(currentDialPosition, rotation);
        }

        _console.WriteLine($"Password: {password}");
    }

    private static int TurnDial(int dialPosition, Rotation rotation)
    {
        var newPosition = rotation.Direction switch
        {
            Direction.Left => dialPosition - rotation.Distance,
            Direction.Right => dialPosition + rotation.Distance,
            _ => throw new NotImplementedException()
        };

        newPosition = (newPosition % UpperDialLimit + UpperDialLimit) % UpperDialLimit;

        return newPosition;
    }

    private List<Rotation> ReadRotations()
    {
        List<Rotation> rotations = [];

        var lines = _reader.ReadLines();
        foreach (var line in lines)
        {
            var direction = line[0] switch
            {
                'L' => Direction.Left,
                'R' => Direction.Right,
                _ => throw new ArgumentException()
            };

            var distance = int.Parse(line[1..]);

            rotations.Add(new(direction, distance));
        }

        return rotations;
    }

    private enum Direction { Left, Right }

    private record struct Rotation(Direction Direction, int Distance);
}

File: Advent of Code.FSharp/Challenge Solutions/Year 2025/ChallengeSolution01.fs
namespace Advent_of_Code.FSharp.Challenge_Solutions.Year_2025

open System
open Advent_of_Code.FSharp.Utilities
open Advent_of_Code.Shared
open Advent_of_Code.Shared.Utilities

module ChallengeSolution01 =

    [<Literal>]
    let UpperDialLimit : int = 100

    [<Literal>]
    let StartingDialPosition : int = 50

    [<Literal>]
    let Zero = 0

    type Rotation =
    | Left of int
    | Right of int

    let distanceOf : Rotation -> int = function
    | Left x -> x
    | Right x -> x

    type Solution(getInput: unit -> string list, writeLine: string -> unit) =

        let readRotations() : Rotation list =
            getInput()
            |> List.map 
                ^<| (Seq.toList
                    >> function
                    | direction::distanceAsString -> 
                        (direction, distanceAsString 
                                    |> List.toArray
                                    |> String
                                    |> int)
                    | _ -> failwithMalformedInput())
            |> List.map 
                ^<| fun (direction, distance) -> 
                    match direction with
                    | 'L' -> Left distance
                    | 'R' -> Right distance
                    | _   -> failwithMalformedInput()

        let turnDial (dialPosition: int) (rotation: Rotation) =
            match rotation with
            | Left distance  -> dialPosition - distance
            | Right distance -> dialPosition + distance
            |> fun dial -> ((dial % UpperDialLimit) + UpperDialLimit) % UpperDialLimit

        member _.SolveFirstPart() =
            
            let rec computePassword (rotations: Rotation list) (dial: int) : int =
                match rotations with
                | [] -> 0
                | rotation::remaining ->
                    let newDialPosition = turnDial dial rotation
                    let password = newDialPosition
                                   |> computePassword remaining

                    match newDialPosition with
                    | Zero -> password + 1
                    | _ -> password

            let rotations = readRotations()
            let password = computePassword rotations StartingDialPosition
            writeLine $"Password: {password}"

        member _.SolveSecondPart() : unit =

            let rec computePassword (rotations: Rotation list) (dial: int) : int =
                match rotations with
                | [] -> 0
                | rotation::remaining ->
                    let distance = distanceOf rotation
                    let fullRotations = distance / UpperDialLimit
                    let remainder = distance - (fullRotations * UpperDialLimit)

                    let hasRotationMovedPastZero = function
                    | Left _ -> dial <= remainder
                    | Right _ -> dial >= UpperDialLimit - remainder

                    let movedPastZeroAddition = if dial <> Zero && hasRotationMovedPastZero rotation then 1 else 0

                    let password = turnDial dial rotation 
                                   |> computePassword remaining

                    password + fullRotations + movedPastZeroAddition

            let rotations = readRotations()
            let password = computePassword rotations StartingDialPosition
            writeLine $"Password: {password}"

    type ChallengeSolution01(console: IConsole, reader: ISolutionReader<ChallengeSolution01>) =
        inherit ChallengeSolution<ChallengeSolution01>(console, reader)

        let solution = Solution(reader.ReadLines >> List.ofArray, console.WriteLine)

        override this.SolveFirstPart() = solution.SolveFirstPart()
        override this.SolveSecondPart() = solution.SolveSecondPart()